### PR TITLE
feature/search

### PR DESCRIPTION
**Pull request by DmytroVasylenko6**

### CODE REVIEW

File: src/js/categories/render-template.js [hunk 1]
```diff
@@ -1,10 +1,12 @@
...
+  return `
   <li class="category-card">
     <a href="/" class="router-link category-link"
-      data-filter=${filter}
+      data-filter="${filter}"
       data-page="1"
       data-limit="12"
-      data-name=${name}
+      data-name="${name}"
+      data-category="${getCategoryName(filter)}"
```
Comment (snegurova) [inline src/js/categories/render-template.js:9]: Category name it's filter

File: src/js/pages/home.js [hunk 2]
```diff
@@ -1,22 +1,21 @@
 import api from '../api';
 import { renderCards } from '../categories/categories-api';
-import { handleClick, addListener } from '../../main';
+import { handleClickFilter } from '../../main';
```
Comment (snegurova) [inline src/js/pages/home.js]: handleClick should be imported from main and modified here if needed

File: src/main.js [hunk 3]
```diff
@@ -18,6 +18,10 @@ export const SEARCH_PARAMS = {
   PAGE: 'page',
   LIMIT: 'limit',
   NAME: 'name',
+  KEYWORD: 'keyword',
+  BODY_PART: 'bodypart',
```
Comment (snegurova) [inline src/main.js:22]: There is FILTERS object for liters, this change brokes existing logic

File: src/main.js [hunk 4]
```diff
@@ -41,42 +45,59 @@ document.addEventListener('DOMContentLoaded', () => {
   handleLocation();
 });
 
-export const getUrl = (event) => {
-  event.preventDefault();
-  const url = new URL(event.currentTarget.href);
+export const getUrl = (stringUrl) => {
```
Comment (snegurova) [inline src/main.js]: this change brokes existing logic

File: src/main.js [hunk 5]
```diff
@@ -41,42 +45,59 @@ document.addEventListener('DOMContentLoaded', () => {
...
+  event.preventDefault();
+  const url = getUrl(event.currentTarget.href);
+  setParams(event.currentTarget.dataset, url);
+  pushState(url);
+};
+
+export const handleClickCategory = (event) => {
+  event.preventDefault();
+  const url = getUrl(window.location.origin);
+  setParams(
```
Comment (snegurova) [inline src/main.js]: this function already sets params, handleClickCategory should be located in categores page or home

File: src/main.js [hunk 6]
```diff
@@ -41,42 +45,59 @@ document.addEventListener('DOMContentLoaded', () => {
...
+export const handleClickFilter = (event) => {
+  event.preventDefault();
+  const url = getUrl(event.currentTarget.href);
+  setParams(event.currentTarget.dataset, url);
+  pushState(url);
+};
+
+export const handleClickCategory = (event) => {
+  event.preventDefault();
+  const url = getUrl(window.location.origin);
```
Comment (snegurova) [inline src/main.js]: we don't need origin

File: src/main.js [hunk 7]
```diff
@@ -94,4 +115,42 @@ export const addListener = (selector, callback) => {
...
+    let previousCall = this.lastCall;
+    this.lastCall = Date.now();
+    if (previousCall && this.lastCall - previousCall <= t) {
+      clearTimeout(this.lastCallTimer);
+    }
+    this.lastCallTimer = setTimeout(() => f(args), t);
+  };
+}
+
+function onSearch(e) {
```
Comment (snegurova) [inline src/main.js]: function on search should take query params and add to existing or replace keyword and put page to 1, after push state is called/ url is not needed

File: src/main.js [hunk 8]
```diff
@@ -41,42 +45,59 @@ document.addEventListener('DOMContentLoaded', () => {
...
     if (searchKeys.includes(key)) {
-      url.searchParams.set(key, dataset[key]);
+      url.searchParams.set(key, queries[key]);
     }
   });
 };
 
 export const handleLocation = async () => {
-  const { pathname, search } = window.location;
+  const { search } = window.location;
```
Comment (snegurova) [inline src/main.js:75]: this change brokes the logic, be carefull when you change functions used all over the project

File: src/js/categories/render-template.js [hunk 9]
```diff
@@ -1,10 +1,12 @@
...
     >
```
Comment (snegurova) [inline src/js/categories/render-template.js:11]: category name is in filter

**Review by snegurova — CHANGES_REQUESTED** [review]

File: src/js/categories/render-template.js (hunk 9 above)
Comment (DmytroVasylenko6) [reply src/js/categories/render-template.js:11]: fixed

File: src/main.js (hunk 8 above)
Comment (DmytroVasylenko6) [reply src/main.js:75]: Just pathname is not used in this function

File: src/main.js (hunk 7 above)
Comment (DmytroVasylenko6) [reply src/main.js]: fixed

File: src/js/pages/home.js (hunk 2 above)
Comment (DmytroVasylenko6) [reply src/js/pages/home.js]: fixed

File: src/main.js (hunk 4 above)
Comment (DmytroVasylenko6) [reply src/main.js]: fixed

File: src/main.js (hunk 5 above)
Comment (DmytroVasylenko6) [reply src/main.js]: fixed

File: src/js/categories/render-template.js (hunk 9 above)
Comment (snegurova) [reply src/js/categories/render-template.js:11]: please remove quotes

File: src/js/pages/exercises.js [hunk 10]
```diff
@@ -3,13 +3,28 @@ import { renderExercises } from '../renderExercises';
...
-  const markup = await renderExercises(exercises);
+const categoryMap = {
+  'body parts': 'bodypart',
+  muscles: 'muscles',
+  equipment: 'equipment',
+};
 
-  cardsContainer.innerHTML = `<ul class="main-exercises">${markup}</ul>`;
+export const getExercises = async (params) => {
+  const exercises = await api.exercises.getExercises(normalizeParams(params));
```
Comment (snegurova) [inline src/js/pages/exercises.js]: it's more update parasms

File: src/js/pages/exercises.js [hunk 11]
```diff
@@ -3,13 +3,28 @@ import { renderExercises } from '../renderExercises';
...
+  )}</ul>`;
 };
 
-export const exercisesElement = document.createElement('div');
-exercisesElement.innerHTML = exercisesTemplate;
+function normalizeParams(params) {
+  const category = params.get('filter')?.toLowerCase() || 'unknown';
+  return {
+    [categoryMap[category] || 'unknown']: params.get('name'),
+    page: params.get('page'),
```
Comment (snegurova) [inline src/js/pages/exercises.js]: params sould be just set, it's not a usual object, it's instance of 
URLSearchParams
--


[URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/set)

Comment (snegurova) [reply src/js/pages/exercises.js]: the function updating params should be called on input change, not when get exercises called, page is always 1, after updating params function pushState is called.

File: src/js/pages/exercises.js (hunk 10 above)
Comment (DmytroVasylenko6) [reply src/js/pages/exercises.js]: This is an api method call. It needs all the parameters

File: src/main.js [hunk 12]
```diff
@@ -59,14 +59,17 @@ export const setParams = (event, url) => {
...
   const urlParams = new URLSearchParams(search);
```
Comment (snegurova) [inline src/main.js:76]: In handleLocation when there is search params getExercises is called

File: src/js/pages/exercises.js (hunk 11 above)
Comment (DmytroVasylenko6) [reply src/js/pages/exercises.js]: We agreed that only the keyword is changed in urlParams

File: src/main.js [hunk 13]
```diff
@@ -94,4 +97,42 @@ export const addListener = (selector, callback) => {
...
+      clearTimeout(this.lastCallTimer);
+    }
+    this.lastCallTimer = setTimeout(() => f(args), t);
+  };
+}
+
+function onSearch(e) {
+  const searchQuery = e.target.value.trim();
+  const url = new URL(window.location.href);
+  url.searchParams.set('keyword', searchQuery);
```
Comment (snegurova) [inline src/main.js]: set page to 1, remove filter, name add filter as separate param with value name

Comment (DmytroVasylenko6) [reply src/main.js]: I agree. Fixed

File: src/js/pages/exercises.js (hunk 11 above)
Comment (DmytroVasylenko6) [reply src/js/pages/exercises.js]: URLSearchParams contain unnecessary parameters for getExercises query

File: src/main.js (hunk 12 above)
Comment (DmytroVasylenko6) [reply src/main.js:76]: getExercises should be called if the name field is present. This is how it works now when clicking on exercises in a specific category

File: src/js/categories/render-template.js (hunk 9 above)
Comment (DmytroVasylenko6) [reply src/js/categories/render-template.js:11]: Снимок экрана 2024-11-10 в 14 46 26
this is how the filter looks like in HTML without quotes

**Review by snegurova — APPROVED** [review]

**Pull request merged**